packages/apollo-engine-reporting/src/plugin.ts

```diff
@@ -0,0 +1,403 @@
+import {
+  GraphQLRequestContext,
+  Logger,
+  InvalidGraphQLRequestError,
+} from 'apollo-server-types';
+import { Headers } from 'apollo-server-env';
+import { GraphQLError } from 'graphql';
+import { Trace } from 'apollo-engine-reporting-protobuf';
+
+import {
+  EngineReportingOptions,
+  GenerateClientInfo,
+  AddTraceArgs,
+  VariableValueOptions,
+  SendValuesBaseOptions,
+} from './agent';
+import { EngineReportingTreeBuilder } from './treeBuilder';
+import { ApolloServerPlugin } from "apollo-server-plugin-base";
+import {
+  PersistedQueryNotFoundError,
+  PersistedQueryNotSupportedError,
+} from 'apollo-server-errors';
+
+type Mutable<T> = { -readonly [P in keyof T]: T[P] };
+
+const clientNameHeaderKey = 'apollographql-client-name';
+const clientReferenceIdHeaderKey = 'apollographql-client-reference-id';
+const clientVersionHeaderKey = 'apollographql-client-version';
+
+// This plugin is instantiated once at server start-up. Each request that the
+// server processes will invoke the `requestDidStart` method which will produce
+// a trace (in protobuf Trace format) for that single request. When the request
+// is done, it passes the Trace back to its associated EngineReportingAgent via
+// the addTrace callback. This class isn't for direct use; its constructor is a
+// private API for communicating with EngineReportingAgent.
+export const plugin = <TContext>(
+  options: EngineReportingOptions<TContext> = Object.create(null),
+  addTrace: (args: AddTraceArgs) => Promise<void>,
+  // schemaHash: string,
+): ApolloServerPlugin<TContext> => {
+  const logger: Logger = options.logger || console;
+  const generateClientInfo: GenerateClientInfo<TContext> =
+    options.generateClientInfo || defaultGenerateClientInfo;
+
+
+  return {
+    requestDidStart(requestContext) {
+      let queryString: string | undefined;
+      const treeBuilder: EngineReportingTreeBuilder =
+        new EngineReportingTreeBuilder({
+          rewriteError: options.rewriteError,
+          logger: requestContext.logger || logger,
+        });
+
+      const metrics: NonNullable<typeof requestContext['metrics']> =
+        ((requestContext as Mutable<typeof requestContext>)
+          .metrics = requestContext.metrics || Object.create(null));
+
+      treeBuilder.startTiming();
+      metrics.startHrTime = treeBuilder.startHrTime;
+
+      if (requestContext.request.http) {
+        treeBuilder.trace.http = new Trace.HTTP({
```

Comment:
Not clear to me why this is here but headers, variables, and client info are in preflight.

Reply:
Agree.  I believe this ended up this way due to an under-informed initial pass on this entire migration prior to the introduction of `ensurePreflight` and some uncertainty about what error phases could result in the various life-cycles not getting triggered.  However, this does not need to remain that way any more and I agree it makes more sense with its companions inside `ensurePreflight`.

Resolved in 0658df5e50283c3985ff3e3717bdff8de2adab9d.